### PR TITLE
Fix Issue 19122 - Postblits and destructors called on members of anonymous unions

### PR DESCRIPTION
**Pull request by RazvanN7**
The problem was that the `overlapped` field of the VarDeclarations associated with anonymous union fields was set in the semantic2 pass, while the field postblit was constructed right at the end of semantic1 for struct declaration.

The patch simply checks if a var declaration is a union scope and if so, sets the overlapped field accordingly.

cc @TurkeyMan

### CODE REVIEW

**Comment by thewilsonator** [comment]
Please target stable!

**Comment by RazvanN7** [comment]
Done.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by thewilsonator** [comment]
Blocked on  https://github.com/dlang/phobos/pull/7047

**Comment by TurkeyMan** [comment]
Will this solve all the `__field` stuff? (postblit, copy, destructor, others?)

**Comment by TurkeyMan** [comment]
Awesome, you made my day! :)

**Comment by TurkeyMan** [comment]
As a bonus task, I think it should be un`@safe` to access any element of a union.

**Comment by thewilsonator** [comment]
A one element should behave then same way as a one element struct and one of the element. For multiple values I think its fine as long as there are no pointers in the union.

**Comment by TurkeyMan** [comment]
@thewilsonator I have no idea what you just said... or what you're talking about?

**Comment by thewilsonator** [comment]
@RazvanN7 this PR doesn't seem to distinguish between anonymous and not-anonymous unions.

**Comment by thewilsonator** [comment]
Looks like you'll need to update `Scope.inunion` to be a `UnionDeclaration` rather than a bool.

**Comment by TurkeyMan** [comment]
What's the difference between anonymous and named unions? This behaviour should surely be the same either way?

**Comment by Geod24** [comment]
> What's the difference between anonymous and named unions?

Anonymous unions are "embedded" in the surrounding member.

```d
struct Foo
{
    union {
        int x;
        Object b;
    }
}

struct Bar
{
    union Named {
        int x;
        Object b;
    }
}

pragma(msg, "Foo: ", Foo.sizeof);
pragma(msg, "Bar: ", Bar.sizeof);
```

On 64 bits:
```
Foo: 8LU
Bar: 1LU
```

**Comment by TurkeyMan** [comment]
RIght... obviously... but that doesn't change whether it's valid for the members of either union to have postblit/destructors and friends called or not, or whether it's valid for more than 1 item to be initalised (the rest should all require `= void` initialisation)...

File: test/runnable/test19122.d
Comment (Geod24) [inline test/runnable/test19122.d]: Those comment are misleading. It should not fail, right ?

**Comment by RazvanN7** [comment]
I think that it doesn't matter of the union is anonymous or not. The phobos failure is legit, you are accessing an overlapped member (even though it is the single union member), it really doesn't matter if the union anonymous or not.

**Comment by Geod24** [comment]
Well the test in this PR currently works with a named union.
As in, I used the following definition for `struct S`:
```d
 struct S
{
    union X
    {
        int i;
        HasDestructor h;
    }
    X field;
}
```

And `dmd -run test.d` (DMD 2.086.0) passed.

**Comment by TurkeyMan** [comment]
Doesn't matter is that works or not... it's blatantly wrong.
```d
struct S
{
    union X
    {
        int i;
        HasDestructor h;
        AlsoHasDestructor j;
    }
    X field;
}
```
You can't have the union members calling lifetime management functions.

**Comment by Geod24** [comment]
@TurkeyMan : Yes and they are *not* called with named unions. Only with anonymous unions. Please re-read my message above.

Alternatively, I suggest using `assert(0)` in test instead of incrementing globals.
So:
```d
struct HasDestructor
{
    ~this() { assert(0); }
    this(this) { assert(0); }
}

struct S
{
    union
    {
        HasDestructor g;
        int i;
        HasDestructor h;
    }
}

void main()
{
    {
        S s;
        s = s;
    }
}
```
Currently: Triggers `assert`. Should not.

```d
struct HasDestructor
{
    ~this() { assert(0); }
    this(this) { assert(0); }
}

struct S
{
    union X // Diff HERE
    {
        HasDestructor g;
        int i;
        HasDestructor h;
    }
    X field; // And HERE
}

void main()
{
    {
        S s;
        s = s;
    }
}
```
Currently (v2.086.0): Does not trigger `assert`. Which is the correct behavior.

**Comment by TurkeyMan** [comment]
Oh okay, sorry. Yes, so named unions work... anonymous unions specifically are broken?
Carry on!

File: test/runnable/test19122.d
Comment (jacob-carlborg) [reply test/runnable/test19122.d]: The test suite is littered with comments like these because the test cases are copy-pasted as is. I’ve mentioned that several times but I’ve been overruled every time.

Comment (Geod24) [reply test/runnable/test19122.d]: I hope we use [`assert(0)`](https://github.com/dlang/dmd/pull/9909#issuecomment-496824330) instead and those lines just disappear.

**Comment by RazvanN7** [comment]
Here's an alternative fix that moves the check inside AnonDeclaration (used for anonymous unions and structs) semantic. The check for anonymous unions cannot be easily done in VarDeclaration semantic since at that point there is no way of retrieving the AnonDeclaration parent (or none that I could think of which is easily implementable).

I've put the if inside the for loop even though it is invariant because I thought it looks nicer than duplicating the for loop.

**Comment by RazvanN7** [comment]
Buildkite failure seems unrelated.

**Comment by Geod24** [comment]
> Buildkite failure seems unrelated.

The failure is:
> runtime/object.d(440): Error: AA key type `ValType` should have `size_t toHash() const nothrow @safe` if `opEquals` defined
> jit/jit.d(116):        while evaluating: `static assert(ValState.sizeof <= 24)`

That's not a spurious failure. [And it is very much related](https://github.com/higgsjs/Higgs/blob/90e2ee898d6ae8ddb9cd8deef294ce80989bd276/source/runtime/object.d#L142-L159).

Which, in this case, is a good thing, because this is probably fixing a bunch of bugs at the same time, by properly marking those fields as overlapped.
But please don't dismiss those CIs results so easily.

File: src/dmd/dsymbolsem.d
Comment (ibuclaw) [inline src/dmd/dsymbolsem.d]: Stylistic nit, give the first if braces?

**Comment by TurkeyMan** [comment]
So, what's doing on here?

**Comment by thewilsonator** [comment]
@RazvanN7 the extracted example that fails:

```
import std.bitmanip;

enum Tag : ubyte
{
    // Note: undef has value zero so that zeroed memory has value undefined
    UNDEF = 0,
    NULL,
    BOOL,
    INT32,
    INT64,
    FLOAT64,
    RAWPTR,
    RETADDR,
    FUNPTR,

    // GC heap pointer tags
    REFPTR,
    OBJECT,
    ARRAY,
    CLOSURE,
    STRING,
    ROPE
}
class ObjShape 
{
    ObjShape[][ValType][wstring] propDefs;
}

struct ValType
{
    // ValType is at most 2 words long
    static assert (ValType.sizeof <= 16);

    static const ValType ANY = ValType();

    union
    {
        /// Shape (null if unknown)
        ObjShape shape;

        /// IR function pointer
        void function() fptr;

        /// Constant value word, if known
        ulong word;
    }
        mixin(bitfields!(

        /// Type tag bits, if known
        Tag, "tag", 4,

        /// Type tag known flag
        bool, "tagKnown", 1,

        /// Shape known flag
        bool, "shapeKnown", 1,

        /// Function pointer known flag (closures and fptrs only)
        bool, "fptrKnown", 1,

        /// Constant value known flag
        bool, "valKnown", 1,

        /// Submaximal flag (overflow check elimination)
        bool, "subMax", 1,

        /// Padding bits
        uint, "", 23
    ));
        // Comparison operator
    bool opEquals(const ValType that) const
    {
        if (this.tagKnown != that.tagKnown)
            return false;

        if (this.shapeKnown != that.shapeKnown)
            return false;

        if (this.fptrKnown != that.fptrKnown)
            return false;

        if (this.valKnown != that.valKnown)
            return false;

        if (this.tagKnown && this.tag != that.tag)
            return false;

        if ((this.shapeKnown || this.fptrKnown || this.valKnown) &&
            (this.word != that.word))
            return false;

        return true;
    }
}

struct ValState
{
    // ValState is at most 3 words long
    static assert (ValState.sizeof <= 24);

    /// Value kind
    enum Kind
    {
        STACK,
        REG,
        CONST
    }

    /// Value type, may be unknown
    ValType type;

    /// Bit field for compact encoding, 32 bits long
    mixin(bitfields!(

        /// Value kind
        Kind, "kind", 2,

        /// Type written to type stack
        bool, "tagWritten", 1,

        /// Local index, or register number
        int, "idx", 24,

        /// Padding bits
        uint, "", 5
    ));
}
```

**Comment by Geod24** [comment]
> So, what's going on here?

Waiting for a green CI. @RazvanN7 can you fix the failure on Higgs' side ?
Also add a test case to DMD.

**Comment by RazvanN7** [comment]
>Waiting for a green CI. @RazvanN7 can you fix the failure on Higgs' side ?

It seems like the fix is to provide a toHash method for the ValType struct, however I don't know what the correct one should be. If I'm not missing something, the builtin hash method [1] for structs seems to simply return the hashing of the field types (a constant?).

[1] https://github.com/dlang/dmd/blob/master/src/dmd/clone.d#L772

**Comment by RazvanN7** [comment]
Pending PR: https://github.com/higgsjs/Higgs/pull/217

**Comment by jacob-carlborg** [comment]
> It seems like the fix is to provide a toHash method for the ValType struct, however I don't know what the correct one should be. I will use the builtin toHash method for structs [1]

So this is a breaking change? Perhaps should not target stable and include a changelog.

**Comment by thewilsonator** [comment]
The behaviour it fixes is horrendously broken and the breaking change it induces is `accepts invalid` i.e. it should have been going that error message all along. A changelog is a good idea though.

**Comment by Geod24** [comment]
I agree with @jacob-carlborg here: If it breaks code, it makes `stable` not so `stable` :)

**Comment by thewilsonator** [comment]
> Pending PR: higgsjs/Higgs#217

Merged. This is green.

**Review by Geod24 — DISMISSED** [review]
- Please target master
- Please add a changelog
- Please apply Iain's review

Then we're good to go IMO

**Comment by RazvanN7** [comment]
@Geod24 Done.

File: src/dmd/dsymbolsem.d
Comment (WalterBright) [reply src/dmd/dsymbolsem.d]: This only superficially works, because decls can be not symbols but can contain symbols, such as:
```
 union { align(1) { int a; long b; } }
```

**Review by WalterBright — DISMISSED** [review]
Needs to take into account nested fields in anonymous unions.

**Comment by RazvanN7** [comment]
@WalterBright Done

**Comment by RazvanN7** [comment]
Ping. I think this is ready to merge.

**Comment by RazvanN7** [comment]
ping @WalterBright

**Comment by thewilsonator** [comment]
Will merge tomorrow morning if no-one else does before that.

**Pull request merged**
